src/osd/PrimaryLogPG.cc

```diff
@@ -3457,24 +3458,121 @@ void PrimaryLogPG::cancel_manifest_ops(bool requeue, vector<ceph_tid_t> *tids)
   }
 }
 
-void PrimaryLogPG::refcount_manifest(ObjectContextRef obc, object_locator_t oloc, hobject_t soid,
-                                     SnapContext snapc, bool get, RefCountCallback *cb, uint64_t offset)
+void PrimaryLogPG::dec_refcount(ObjectContextRef obc, const object_info_t& oi, 
+				const object_ref_delta_t& refs)
+{
+  for (auto p = refs.begin(); p != refs.end(); ++p) {
+    dout(10) << __func__ << ": decrement reference on offset oid: " << p->first << dendl;
+    object_locator_t target_oloc(p->first);
+    refcount_manifest(obc, obc->obs.oi.soid, target_oloc, p->first, 
+		      SnapContext(), refcount_t::DECREMENT_REF, NULL);
+  }
+}
+
+
+void PrimaryLogPG::dec_all_refcount_head_manifest(object_info_t& oi, OpContext* ctx, const hobject_t &coid)
+{
+  SnapSetContext* ssc = ctx->obc->ssc;
+  ceph_assert(oi.has_manifest());
+  ceph_assert(oi.soid.is_head());
+  // has snapshot
+  if (ssc && ssc->snapset.clones.size() > 0) {
+    ceph_assert(ssc);
+    dout(15) << __func__ <<  " has snapset " << dendl;
+    object_ref_delta_t refs;
+    ceph_assert(oi.manifest.is_chunked());
+    ceph_assert(!oi.manifest.is_redirect());
+
+    SnapSet& snapset = ssc->snapset;
+
+    // check adjacent clones
+    auto s = std::find(snapset.clones.begin(), snapset.clones.end(), coid.snap);
+    int index = std::distance(snapset.clones.begin(), s);
+    snapid_t l = *snapset.clones.begin(), g = CEPH_NOSNAP;
+    object_manifest_t* manifest_g = nullptr, * manifest_l = nullptr;
+    hobject_t t_oid = coid;
+
+    auto get_context = [this](const hobject_t & oid) 
+      -> ObjectContextRef {
+      ObjectContextRef obc = get_object_context(oid, false, NULL);
+      ceph_assert(obc);
+      ceph_assert(obc->obs.oi.has_manifest());
+      ceph_assert(obc->obs.oi.manifest.is_chunked());
+      return obc;
+    };
+
+    if (index) {
+      l = snapset.clones[index - 1];
+      t_oid.snap = l;
+      manifest_l = &get_context(t_oid)->obs.oi.manifest;
+    } 
+
+    if (*s != snapset.clones.back() && s != snapset.clones.end()) {
```

Comment:
*s is undefined behavior if s == snapset.clones.end() -- you need to reverse these conditions.  Also, snapset.clones.back() is undefined if snapset.clones.size() == 0.